File: solutions/cpp/kindergarten-garden/2/kindergarten_garden.cpp
#include "kindergarten_garden.h"
#include <algorithm>
#include <iterator>
#include <cassert>
#include <map>
#include <stdexcept>
#include <vector>
namespace kindergarten_garden {

// TODO: add your solution here

    const std::vector<std::string> STUDENTS{
        "Alice",
        "Bob",
        "Charlie",
        "David",
        "Eve",
        "Fred",
        "Ginny",
        "Harriet",
        "Ileana",
        "Joseph",
        "Kincaid",
        "Larry"
    };

    const std::map<char, Plants> PLANT_CODE{
        {'C', Plants::clover},
        {'G', Plants::grass},
        {'V', Plants::violets},
        {'R', Plants::radishes}
    };

    std::array<Plants, 4> plants(const std::string& plant_assignment, const std::string& student) {
        const size_t pos = plant_assignment.find("\n");
        assert(pos != std::string::npos);
        std::string row[2] = {
            plant_assignment.substr(0, pos),
            plant_assignment.substr(pos + 1)
        };
        auto student_index = static_cast<size_t>(std::distance(STUDENTS.begin(), std::find(STUDENTS.begin(), STUDENTS.end(), student)));
        if (student_index == STUDENTS.size())
            throw std::domain_error("unknown student");
        
        auto plant_index = student_index * 2;
        if (plant_index >= row[0].length() || plant_index >= row[1].length())
            throw std::domain_error("invalid argument value");

        return {
            PLANT_CODE.at(row[0][plant_index]), PLANT_CODE.at(row[0][plant_index + 1]),
            PLANT_CODE.at(row[1][plant_index]), PLANT_CODE.at(row[1][plant_index + 1])
        };
    }

}  // namespace kindergarten_garden

File: solutions/python/kindergarten-garden/2/kindergarten_garden.py
class Garden:
    PLANTS_CODE = {
        "G": "Grass",
        "C": "Clover",
        "R": "Radishes",
        "V": "Violets"
    }

    DEFAULT_STUDENTS = ["Alice", "Bob", "Charlie", "David", "Eve", "Fred",
                        "Ginny", "Harriet", "Ileana", "Joseph", "Kincaid", "Larry"]

    def __init__(self, diagram: str, students: list[str] = None):
        self.diagram = diagram
        if students is None:
            students = Garden.DEFAULT_STUDENTS
        self.students = sorted(students)
        student_count = len(self.students)
        self.diagram: list[str] = diagram.split()
        diagram_len = list(map(len, self.diagram))
        if not all(map(lambda x: x == diagram_len[0], diagram_len)):
            raise ValueError("Invalid diagram format")

    def plants(self, student: str) -> list[str]:
        if student not in self.students:
            raise ValueError(f"'{student}' not found")
        index = self.students.index(student) * 2
        return [Garden.PLANTS_CODE.get(line[n], "unknown")
                for line in self.diagram
                for n in (index, index+1)]
